crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,345 @@
+        this
+    }
+
+    /// Requests the suggested tip for the latest block.
+    pub(crate) async fn suggest_tip(&self) -> Result<U256> {
+        let (response_tx, rx) = oneshot::channel();
+        let _ = self.to_service.send(GasPriceAction::SuggestMaxPriorityFee { response_tx });
+        // TODO: not sure if this is the right error, just modelling after the cache
+        rx.await.map_err(|_| ProviderError::GasPriceOracleServiceUnavailable)?
+    }
+}
+
+/// A task that manages gas price estimates.
+///
+/// Caution: The channel for the data is _unbounded_ it is assumed that this is mainly used by the
+/// [EthApi](crate::EthApi) which is typically invoked by the RPC server, which already uses permits
+/// to limit concurrent requests.
+#[must_use = "Type does nothing unless spawned"]
+pub(crate) struct GasPriceOracleService<Client, Tasks> {
+    /// The type used to subscribe to block events and get block info
+    client: Client,
+    /// The config for the oracle
+    oracle_config: GasPriceOracleConfig,
+    /// The latest calculated price and its block hash
+    last_price: GasPriceOracleResult,
+    /// Sender half of the action channel.
+    action_tx: UnboundedSender<GasPriceAction>,
+    /// Receiver half of the action channel.
+    action_rx: UnboundedReceiverStream<GasPriceAction>,
+    /// The type that's used to spawn tasks that do the actual work
+    action_task_spawner: Tasks,
+}
+
+impl<Client, Tasks> GasPriceOracleService<Client, Tasks>
+where
+    Client: BlockProvider + Clone + Unpin + 'static,
+    Tasks: TaskSpawner + Clone + 'static,
+{
+    // TODO: concurrent calls for get_block_values like in the geth impl
+    fn suggest_tip_cap(&mut self) -> Result<U256> {
+        // TODO: we really shouldn't have this be None, but let's get rid of these expects?
+        let header_hash = self
+            .client
+            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
+            .expect("a latest header always exists");
+        let header = self
+            .client
+            .header_by_hash_or_number(BlockHashOrNumber::Hash(header_hash))?
+            .ok_or(ProviderError::BlockHash { block_hash: header_hash })?;
+
+        // if we have stored a last price, then we check whether or not it was for the same head
+        if self.last_price.block_hash == header_hash {
+            return Ok(self.last_price.price)
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
```

Comment:
this looks wasteful to me to fetch the entire history range again.

Reply:
`get_block_values` should use the `EthStateCache` now